libsolidity/analysis/ReferencesResolver.cpp

```diff
@@ -301,109 +301,66 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
-				);
-				if (contract.isLibrary())
+				string allowedStorageLocationsStr;
+				string locStr;
+				for (auto const loc : allowedStorageLocations)
 				{
-					if (varLoc == Location::Memory)
-						fatalTypeError(_variable.location(),
-							"Location has to be calldata or storage for external "
-							"library functions (remove the \"memory\" keyword)."
-						);
+					if (!allowedStorageLocationsStr.empty())
+						allowedStorageLocationsStr += " or ";
+					if (loc == Location::Default)
+						locStr = "none";
+					else
+						locStr = "\"" + _variable.locationToString(loc) + "\"";
+					allowedStorageLocationsStr += locStr;
 				}
-				else
+				string errorString = "Storage location must be " + allowedStorageLocationsStr;
+				if (_variable.isCallableParameter())
 				{
-					// force location of external function parameters (not return) to calldata
-					if (varLoc != Location::CallData && varLoc != Location::Default)
-						fatalTypeError(_variable.location(),
-							"Location has to be calldata for external functions "
-							"(remove the \"memory\" or \"storage\" keyword)."
-						);
+					auto const &varScope = dynamic_cast<Declaration const &>(*_variable.scope());
+					string constant = _variable.isConstant() ? " constant" : "";
+					errorString += " for" + constant + " parameter in ";
+					errorString += varScope.visibilityToString(varScope.visibility()) + " function";
 				}
-				if (varLoc == Location::Default)
-					typeLoc = DataLocation::CallData;
-				else
-					typeLoc = varLoc == Location::Memory ? DataLocation::Memory : DataLocation::Storage;
-			}
-			else if (_variable.isCallableParameter() && dynamic_cast<Declaration const&>(*_variable.scope()).isPublic())
-			{
-				auto const& contract = dynamic_cast<ContractDefinition const&>(
-					*dynamic_cast<Declaration const&>(*_variable.scope()).scope()
-				);
-				// force locations of public or external function (return) parameters to memory
-				if (varLoc != Location::Memory && varLoc != Location::Default && !contract.isLibrary())
-					fatalTypeError(_variable.location(),
-						"Location has to be memory for publicly visible functions "
-						"(remove the \"storage\" or \"calldata\" keyword)."
-					);
-				if (varLoc == Location::Default || !contract.isLibrary())
-					typeLoc = DataLocation::Memory;
 				else
-				{
-					if (varLoc == Location::CallData)
-						fatalTypeError(_variable.location(),
-							"Location cannot be calldata for non-external functions "
-							"(remove the \"calldata\" keyword)."
-						);
-					typeLoc = varLoc == Location::Memory ? DataLocation::Memory : DataLocation::Storage;
-				}
+					errorString += " for variable";
+				string varLocStr = _variable.locationToString(varLoc);
+				varLocStr = varLocStr == "none" ? "none" : "\"" + varLocStr + "\"";
+				errorString += ", but " + varLocStr + " was given.";
+				fatalTypeError(_variable.location(), errorString);
 			}
+
+			// Find correct data location.
```

Comment:
If possible, I would like to have an assertion here that checks that we assign here is compatible with `varLoc`, i.e. either typeLoc matches varLoc or varLoc is "none".